credentials/tls/certprovider/distributor.go

```diff
@@ -0,0 +1,108 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package certprovider
+
+import (
+	"context"
+	"sync"
+
+	"google.golang.org/grpc/internal/grpcsync"
+)
+
+// Distributor makes it easy for provider implementations to furnish new key
+// materials by handling synchronization between the producer and consumers of
+// the key material.
+//
+// Provider implementations must embed this type into themselves and:
+// - Whenever they have new key material, they should invoke the Set() method.
+// - When users of the provider call KeyMaterial(), it will be handled by the
+//   distributor which will return the most up-to-date key material furnished
+//   by the provider.
+// - When users of the provider call Close(), the channel returned by the
+//   Done() method will be closed. So, provider implementations can select on
+//   the channel returned by Done() to perform cleanup work, or override
+//   Close(), in which case they must invoke Distributor.Close() when the
+//   provider is closed.
+type Distributor struct {
+	mu sync.Mutex
+	km *KeyMaterial
+
+	ready  *grpcsync.Event
+	closed *grpcsync.Event
+}
+
+// NewDistributor returns a new Distributor.
+func NewDistributor() *Distributor {
+	return &Distributor{
+		ready:  grpcsync.NewEvent(),
+		closed: grpcsync.NewEvent(),
+	}
+}
+
+// Set updates the key material in the distributor with km.
+func (d *Distributor) Set(km *KeyMaterial) {
+	d.mu.Lock()
+	d.km = km
+	d.ready.Fire()
+	d.mu.Unlock()
+}
+
+// KeyMaterial returns the most recent key material provided to the distributor.
+// If no key material was provided to the distributor at the time of this call,
+// it will block until the deadline on the context expires or fresh key material
+// arrives.
+func (d *Distributor) KeyMaterial(ctx context.Context, opts KeyMaterialOptions) (*KeyMaterial, error) {
+	if d.closed.HasFired() {
+		return nil, errProviderClosed
+	}
+
+	if d.ready.HasFired() {
```

Comment:
Do we need this if? Is this faster than channel receive?
(We need the close check, in case ready is also closed, to avoid the race).

Reply:
I noticed this, too.  This will allow us to have a consistent result when the context is canceled upon entry to the function and the distributor is not empty.  Otherwise it will be 50/50 between the two results.

Reply:
The `HasFired` call would be faster than the channel receive as the former is an atomic load. But my intention was not to optimize for speed. My intention was to make it explicit that if we have key_materials (`ready` has fired), then use it immediately instead of selecting on a bunch of channels.

Do you think it is better to get rid of this explicit check, and let the select handle everything?

Reply:
I personally would prefer the explicit check, to have deterministic behavior.